File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
from collections import deque

queue = deque()

def bfs(numbers, target):
    ans = 0
    queue.append((numbers[0], 1))
    queue.append((-numbers[0], 1))
    while queue:
        x, i = queue.popleft()
        if i >= len(numbers):
            if x == target:
                ans += 1
            continue
        queue.append((x+numbers[i], i+1))
        queue.append((x-numbers[i], i+1))

    return ans

def solution(numbers, target):
    cnt = bfs(numbers, target)
    return cnt

numbers = [1,1,1,1,1]
target = 3
print(solution(numbers, target))

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
import java.util.*;

class Solution {
    public int solution(int[] numbers, int target) {
        int answer = 0;
        
        Queue<int[]> queue = new ArrayDeque<>();
        queue.add(new int[]{0, numbers[0]});
        queue.add(new int[]{0, -numbers[0]});
        
        while (!queue.isEmpty()) {
            int[] p = queue.poll();
            int idx = p[0], sum = p[1];
            
            if (sum == target && (idx == numbers.length-1)) {
                answer++;
                continue;
            }
            
            if (idx == numbers.length-1) {
                continue;
            }
            
            queue.add(new int[]{idx+1, sum+numbers[idx+1]});
            queue.add(new int[]{idx+1, sum-numbers[idx+1]});
        }
        
        return answer;
    }
}